Add RFC 105 text: Add and use safe path manipulation functions (#11640)
OSGeo/gdal: doc/source/development/rfc/rfc105_safe_path_manipulation_functions.rst

.. _rfc-105:

===================================================================
RFC 105: Add and use safe path manipulation functions
===================================================================

============== =============================================
Author:        Even Rouault
Contact:       even.rouault @ spatialys.com
Started:       2025-01-11

> Status:        Adopte, implemented in 3.11

Author: ```suggestion
Status:        Adopted, implemented in 3.11
```

.. _rfc-105:

===================================================================
RFC 105: Add and use safe path manipulation functions
===================================================================

============== =============================================
Author:        Even Rouault
Contact:       even.rouault @ spatialys.com
Started:       2025-01-11
Status:        Draft
Target:        GDAL 3.11
============== =============================================

Summary
-------

This RFC adds safe versions, for use by C++ code, of all functions of
cpl_path.cpp (such as CPLGetPath(), CPLGetDirname(), CPLGetBasename(), CPLGetExtension(),

> CPLGetFormFilename(), CPLGetFormCIFilename(), etc.), that returns a result stored

Reviewer: ? should be CPLFormFilename() and CPLFormCIFilename() (no "Get")

Author: indeed, thanks! Fixed

> CPLGetFormFilename(), CPLGetFormCIFilename(), etc.), that returns a result stored

Author: ```suggestion
CPLFormFilename(), CPLFormCIFilename(), etc.), that returns a result stored
```

in more or less ephemeral storage, to avoid potential security issues related
to their mis-use. It also covers converting most of the code base to the safer
alternatives.

Motivation
----------

Above mentioned functions return a ``const char *`` string, which is owned by

> a thread-locale rotating buffer of 10 strings of up to 2048 bytes each, managed

Reviewer 2: Nit:

```suggestion
a thread-local rotating buffer of 10 strings of up to 2048 bytes each, managed
```

cases "just" cause malfunction, in bad cases, reading outside the bounds of
buffers, and perhaps even worse when unlucky.

This risk of such mis-use has been identified for long, and the documentation of
those functions states that the result they return is short-lived, and that
it must be stored in a longer-living storage when not immediately used.

Thus, patterns like the following is the recommended practice:

.. code-block:: c++

        const std::string osExt = CPLGetExtension(mainFile);
        for (int i = 0; i < 10; ++i )
        {
            do_something(std::string(CPLGetExtension(auxiliaryFiles[i])).c_str());
        }
        do_something_else(osExt.c_str());

However, experience has shown that it is extremely easy to forget that, and over
the years, this has caused several heap buffer overflows bugs detected by
oss-fuzz, and probably latent ones still uncovered.
The later in date is https://issues.oss-fuzz.com/issues/388868487 , fixed
by https://github.com/OSGeo/gdal/pull/11638 , and shows how subtle those bugs
can be. In that instance, the issue is due to complex chains of calls where we

> end up using more than 10 instances of the thread-locale storage.

Reviewer 2: ```suggestion
end up using more than 10 instances of the thread-local storage.
```

Implementation
--------------

Rather to react on a case by case when fuzzers found issues, we want a
systematic approach to suppress their root case:

- Start from the implementation of the CPLXXXX() functions, copy them to
  C++ CPLXXXXSafe() functions that return a ``std::string`` instead of a ``const char *``,

>   and modify them to no longer use thread-locale rotating buffer of strings,

Reviewer 2: ```suggestion
  and modify them to no longer use thread-local rotating buffer of strings,
```

  but rely on std::string natural concatenation routines instead
  of the convoluted C concatenation logic, resulting in clearer code.

- make the now superseded CPLXXXX() functions call the Safe versions, and store

>   the result in the thread-locale rotating buffer of strings. So those functions

Reviewer 2: ```suggestion
  the result in the thread-local rotating buffer of strings. So those functions
```

  we can in some cases add a std::string member variable to store the result
  of the safe methods and return its C-string.

Impact on performance
---------------------

There might be a theoretical impact on performance due to dynamic memory
allocation done by temporary ``std::string`` allocations, although normally
those uses occur in the identification and open part of drivers, and not in
performance critical loops. As the identification part is still critical, and
its main use if to get the filename extension, we extend the GDALOpenInfo class,
so it stores the filename extension as a member variable, which will save tens
of drivers from the need of calling CPLGetExtensionSafe(), as well as a
convenience `GDALOpenInfo::IsExtensionEqualToCI(const char* pszExtension)` method.
So, all in all, the performance impact of those changes is thought/hoped to be
in the hardly measurable category.

Out-of-scope (for the candidate implementation related to this RFC)
-------------------------------------------------------------------

We have exactly the same issue with ``const char* CPLSPrintf(const char* pszFmt, ....)``

> which uses its own thread-locale rotating buffer of 10 strings of size up to

Reviewer 2: ```suggestion
which uses its own thread-local rotating buffer of 10 strings of size up to
```
